go/libraries/doltcore/merge/constraint_violations.go

```diff
@@ -0,0 +1,507 @@
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		postChildCVMap, err = types.NewMap(ctx, postChildTable.ValueReadWriter())
+		if err != nil {
+			return nil, err
+		}
+	}
+	postChildCVMapEditor := postChildCVMap.Edit()
+	postChildTableEditor, err := editor.NewTableEditor(ctx, postChildTable, postChildSch, foreignKey.TableName)
+	if err != nil {
+		return nil, err
+	}
+
+	differ := diff.NewRowDiffer(ctx, preParentSch, postParentSch, 1024)
+	defer differ.Close()
+	differ.Start(ctx, preParentData, postParentData)
+	for {
+		diffSlice, hasMore, err := differ.GetDiffs(1, 10*time.Second)
+		if err != nil {
+			return nil, err
+		}
+		if len(diffSlice) != 1 {
+			if hasMore {
+				return nil, fmt.Errorf("no diff returned but should have errored earlier")
+			}
+			break
+		}
+		rowDiff := diffSlice[0]
+		switch rowDiff.ChangeType {
+		case types.DiffChangeRemoved, types.DiffChangeModified:
+			postParentRow, err := row.FromNoms(postParentSch, rowDiff.KeyValue.(types.Tuple), rowDiff.OldValue.(types.Tuple))
+			if err != nil {
+				return nil, err
+			}
+			hasNulls := false
+			for _, tag := range postParentIndexTags {
+				if postParentRowEntry, ok := postParentRow.GetColVal(tag); !ok || types.IsNull(postParentRowEntry) {
+					hasNulls = true
+					break
+				}
+			}
+			if hasNulls {
+				continue
+			}
+
+			postParentIndexPartialKey, err := row.ReduceToIndexPartialKey(postParentIndex, postParentRow)
+			if err != nil {
+				return nil, err
+			}
+			var mapIter table.TableReadCloser = noms.NewNomsRangeReader(postParentIndexSch, postParentIndexData,
+				[]*noms.ReadRange{{Start: postParentIndexPartialKey, Inclusive: true, Reverse: false, Check: func(tuple types.Tuple) (bool, error) {
+					return tuple.StartsWith(postParentIndexPartialKey), nil
+				}}})
+			//TODO: deal with this defer better
+			defer mapIter.Close(ctx)
+			if _, err := mapIter.ReadRow(ctx); err == nil {
+				// If the parent has more rows that satisfy the partial key then we choose to do nothing
+				continue
+			} else if err != io.EOF {
+				return nil, err
+			}
+			postParentIndexPartialKeySlice, err := postParentIndexPartialKey.AsSlice()
+			if err != nil {
+				return nil, err
+			}
+			for i := 0; i < len(postChildIndexTags); i++ {
+				postParentIndexPartialKeySlice[2*i] = types.Uint(postChildIndexTags[i])
+			}
+			postChildIndexPartialKey, err := types.NewTuple(postChildTable.Format(), postParentIndexPartialKeySlice...)
+			if err != nil {
+				return nil, err
+			}
+			mapIter = noms.NewNomsRangeReader(postChildIndexSch, postChildIndexData,
+				[]*noms.ReadRange{{Start: postChildIndexPartialKey, Inclusive: true, Reverse: false, Check: func(tuple types.Tuple) (bool, error) {
+					return tuple.StartsWith(postChildIndexPartialKey), nil
+				}}})
+			//TODO: deal with this defer better
+			defer mapIter.Close(ctx)
+			var postChildIndexRow row.Row
+			for postChildIndexRow, err = mapIter.ReadRow(ctx); err == nil; postChildIndexRow, err = mapIter.ReadRow(ctx) {
```

Comment:
extract this into a method to fix the defer close issue?  Also this function is a beast so could also help with clarity to pull this out.